### PR TITLE
Add more supervisory parens

### LINKED ISSUE #187: Removes parentheses in expressions with mixed operators
Before:

```js
const foo = (a && b) || c || d;
```

After:

```js
const foo = a && b || c || d;
```

The same goes for arithmetic and bitwise operators.

Relevant ESLint rule: http://eslint.org/docs/rules/no-mixed-operators.

### PR DESCRIPTION
**Pull request by azz**
With this change:

1. Any use of `%` with `/`, `*`, or `%` will add parens on the LHS.
2. Any combination of `==`, `!=`, `===`, `!==` will add parens on the LHS.
3. `x ** y ** z` will become `x ** (y ** z)`, as exponentiation is right-associative.

Closes #187

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -4089,11 +4089,7 @@ function printBinaryishExpressions(
     // precedence level and should be treated as a separate group, so
     // print them normally. (This doesn't hold for the `**` operator,
     // which is unique in that it is right-associative.)
-    if (
-      util.getPrecedence(node.left.operator) ===
-        util.getPrecedence(node.operator) &&
-      node.operator !== "**"
-    ) {
+    if (util.shouldFlatten(node.operator, node.left.operator)) {
```
Comment (azz) [inline src/printer.js]: Changing this was required due to `needsParens` not actually being called on the to-be-flattened expression.

**Comment by lydell** [comment]
What about bitwise operators?

**Comment by azz** [comment]
Bitwise operators are already parenthesised.

**Comment by vjeux** [comment]
Awesome! So good to close a 3 digits issue :)

**Comment by azz** [comment]
@lydell

Bitwise operators _should_ already be parenthesised. Most of them have different precedences, but it looks like shift operators are not parenthesised as expected.

https://prettier.io/playground/#%7B%22content%22%3A%22%5CnA%3C%3Cb%3E%3E%3Ec%3B%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D

I'll fix this tomorrow.

**Pull request merged**